src/s3path.jl

```diff
@@ -638,3 +638,18 @@ function FilePathsBase.mktmpdir(parent::S3Path)
     fp = parent / string(uuid4(), "/")
     return mkdir(fp)
 end
+
+const S3PATH_ARROW_NAME = Symbol("JuliaLang.AWSS3.S3Path")
+ArrowTypes.arrowname(::Type{<:S3Path}) = S3PATH_ARROW_NAME
+ArrowTypes.ArrowType(::Type{<:S3Path}) = String
+ArrowTypes.JuliaType(::Val{S3PATH_ARROW_NAME}, ::Any) = S3Path
+ArrowTypes.fromarrow(::Type{<:S3Path}, uri_string) = S3Path(uri_string)
+
+function ArrowTypes.toarrow(path::S3Path)
+    isnothing(path.config) || throw(
+        ArgumentError(
+            "`path.config` must be `nothing` to serialize `path::S3Path` to Arrow"
+         ),
+    )
+    return string(path)
+end
```

Comment:
```suggestion
ArrowTypes.fromarrow(::Type{<:S3Path}, uri_string) = S3Path(uri_string)
ArrowTypes.toarrow(path::S3Path{Nothing}) = string(path)

function ArrowTypes.toarrow(::S3Path)
    throw(ArgumentError("`S3Path` config must be `nothing` to serialize to Arrow"))
end
```

Reply:
I strongly prefer not to dispatch on type parameters for behavior like this, since it's not correct for the full domain of the type parameter (which is a `Union`, IIUC). 

The existing implementation should be just as performant and is more correct / "to-the-point"